[Migration] document sfx configuration settings equivalents (#124)
signalfx/splunk-otel-dotnet: docs/migrating.md

# Migrating

> This doc outlines configuration settings available in [signalfx-dotnet-tracing](https://github.com/signalfx/signalfx-dotnet-tracing/)
>  instrumentation, and their equivalents available in
>  Splunk distribution of OpenTelemetry .NET.

Reviewer: ```suggestion
## Configuration

This section outlines configuration settings available in
[SignalFx Instrumentation for .NET](https://github.com/signalfx/signalfx-dotnet-tracing/),
and their equivalents in
Splunk distribution of OpenTelemetry .NET.
```

## Configuration

This section outlines configuration settings available in
[SignalFx Instrumentation for .NET](https://github.com/signalfx/signalfx-dotnet-tracing/),
and their equivalents in
Splunk distribution of OpenTelemetry .NET.

## Main settings

The following settings are common to most instrumentation scenarios:

| Setting                   | Description                                                                           | OpenTelemetry equivalent                       |
|---------------------------|---------------------------------------------------------------------------------------|------------------------------------------------|
| `SIGNALFX_ENV`            | The value for the `deployment.environment` tag added to every span.                   | Should be set using `OTEL_RESOURCE_ATTRIBUTES` |
| `SIGNALFX_SERVICE_NAME`   | The name of the application or service.                                               | `OTEL_SERVICE_NAME`                            |
| `SIGNALFX_VERSION`        | The version of the application. When set, it populates the `version` tag on spans.    | Should be set using `OTEL_RESOURCE_ATTRIBUTES` |

## Global management settings

> | Setting                               | Description                                                                                                                                          | Default                                           | OpenTelemetry equivalent                                                                                   |

Author: specified in https://github.com/signalfx/splunk-otel-dotnet/pull/124/commits/8b847f975a21de22ff2ec68b8a96cea358758dcb

## Global instrumentation settings

| Setting                              | Description                                                                                                                                                                                                                                | Default   | OpenTelemetry equivalent                                                                                              |
|--------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|-----------------------------------------------------------------------------------------------------------------------|

> | `SIGNALFX_DISABLED_INTEGRATIONS`     | Comma-separated list of disabled library instrumentations.                                                                                                                                                                                 |           | `OTEL_DOTNET_AUTO_TRACES_DISABLED_INSTRUMENTATIONS`                                                                   |

Reviewer 2: I think that information about supported integrations has to be included. The value is not 1:1 mapping.

Reviewer: Or at least mention where it could be found.

| `SIGNALFX_ACCESS_TOKEN`                  | Splunk Observability Cloud access token for your organization. It enables sending telemetry directly to the Splunk Observability Cloud ingest endpoint.                                               |                                      | `SPLUNK_ACCESS_TOKEN`                                               |
| `SIGNALFX_REALM`                         | Your Splunk Observability Cloud realm. To find your realm, open Splunk Observability Cloud, click Settings, then click on your username.                                                              | `none` (local collector)             | `SPLUNK_REALM`                                                      |
| `SIGNALFX_ENDPOINT_URL`                  | The URL to where trace exporters send traces. Overrides `SIGNALFX_REALM` configuration for the traces ingestion endpoint.                                                                             | `http://localhost:9411/api/v2/spans` | `OTEL_EXPORTER_OTLP_ENDPOINT`/`OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`  |

> | `SIGNALFX_METRICS_ENDPOINT_URL`          | The URL to where metric exporters send metrics. Overrides `SIGNALFX_REALM` configuration for the metrics ingestion endpoint.                                                                          | `http://localhost:9943/v2/datapoint` | `OTEL_EXPORTER_OTLP_ENDPOINT`/`OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` |

Reviewer 2: As I remember `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` nor `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT` are not yet supported by OTel .NET SDK

| `SIGNALFX_TRACE_PARTIAL_FLUSH_ENABLED`   | Enable to export traces that contain a minimum number of closed spans, as defined by `SIGNALFX_TRACE_PARTIAL_FLUSH_MIN_SPANS`.                                                                        | `false`                              | none                                                                |
| `SIGNALFX_TRACE_PARTIAL_FLUSH_MIN_SPANS` | Minimum number of closed spans in a trace before it's exported. The default value is ``500``. Requires the value of the ``SIGNALFX_TRACE_PARTIAL_FLUSH_ENABLED`` environment variable to be ``true``. | `500`                                | none                                                                |
| `SIGNALFX_TRACE_BUFFER_SIZE`             | The size of the trace exporter buffer, expressed as the number of traces.                                                                                                                             | `1000`                               | `OTEL_BSP_MAX_QUEUE_SIZE` (default: 2048)                           |

## Trace propagation settings

> | Setting                | Description                                                                                                                                                   | Default   | Otel                                                                               |

Reviewer 2: Otel? ;)

Author: fixed in https://github.com/signalfx/splunk-otel-dotnet/pull/124/commits/8b847f975a21de22ff2ec68b8a96cea358758dcb

|------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|------------------------------------------------------------------------------------|

> | `SIGNALFX_PROPAGATORS` | Comma-separated list of the propagators for the tracer. Available propagators are: `B3`, `W3C`. The Tracer will try to execute extraction in the given order. | `B3,W3C`  | `OTEL_PROPAGATORS`(default: `tracecontext`, `baggage`, available: `b3multi`, `b3`) |

Reviewer 2: other available?

b3 should not be listed here: https://github.com/signalfx/gdi-specification/blob/6d04201f3472c83792666d1477aa770027c9a069/specification/configuration.md?plain=1#L214-L216

Lack of information that B3 should be changed to b3multi and W3C to `tracecontext,baggage`.
Lack of information that if the user want to have same default as previous he should make additional Information.

Author: The set of values available comes from [here](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation/Configuration/Propagator.cs)
Below is the mapping between `OTEL_PROPAGATORS` and `SIGNALFX_PROPAGATORS` values (comes from [here](https://github.com/signalfx/signalfx-dotnet-tracing/blob/main/docs/advanced-config.md#trace-propagation-settings)).
Not sure why b3 should not be listed - it is supported as evident in the code. The spec states what must be supported, not that b3 mustn't be supported.

Author: additional info about configuration required to keep the same propagators configured after migrating from signalfx added in https://github.com/signalfx/splunk-otel-dotnet/pull/124/commits/8b847f975a21de22ff2ec68b8a96cea358758dcb

| `SIGNALFX_HTTP_SERVER_ERROR_STATUSES`                  | Comma-separated list of HTTP server response statuses for which the spans are set as errors, for example: `300, 400-599`.                                                                                                     | `500-599`                           | none                                                     |
| `SIGNALFX_INSTRUMENTATION_ELASTICSEARCH_TAG_QUERIES`   | Enable the tagging of an Elasticsearch command PostData as `db.statement`. It might introduce overhead for direct streaming users.                                                                                            | `true`                              | none                                                     |
| `SIGNALFX_INSTRUMENTATION_MONGODB_TAG_COMMANDS`        | Enable the tagging of a Mongo command BsonDocument as `db.statement`.                                                                                                                                                         | `true`                              | Not configurable using environment variable.             |

> | `SIGNALFX_INSTRUMENTATION_REDIS_TAG_COMMANDS`          | Enable the tagging of a Redis command as `db.statement`.                                                                                                                                                                      | `true`                              | Not configurable using environment variable.             |

Reviewer 2: I would put here just none.
We do now want to ask users to implement plugins.

| `SIGNALFX_LOGS_INJECTION`                              | Enable to inject trace IDs, span IDs, service name, and environment into logs.                                                                                                                                                | `false`                             | Logs correlated if `Microsoft.Extensions.Logging` used.  |
| `SIGNALFX_TRACE_DELAY_WCF_INSTRUMENTATION_ENABLED`     | Enable the updated WCF instrumentation that delays execution until later in the WCF pipeline when the WCF server exception handling is established.                                                                           | `false`                             | none                                                     |

> | `SIGNALFX_TRACE_HEADER_TAGS`                           | Comma-separated map of HTTP header keys to tag names, automatically applied as tags on traces.                                                                                                                                | `"x-my-header:my-tag,header2:tag2"` | none                                                     |

Reviewer 2: What's about OTEL_RESOURCE_ATTRIBUTES?

Author: I don't think `OTEL_RESOURCE_ATTRIBUTES` should be mentioned here.
Value for attributes added with `OTEL_RESOURCE_ATTRIBUTES` are static, with `SIGNALFX_TRACE_HEADER_TAGS` value for a tag is extracted from a header.
`OTEL_RESOURCE_ATTRIBUTE` is already mentioned as a equivalent for `SIGNALFX_GLOBAL_TAGS`.

# Migrating

## Configuration

This section outlines configuration settings available in
[SignalFx Instrumentation for .NET](https://github.com/signalfx/signalfx-dotnet-tracing/),
and their equivalents in
Splunk distribution of OpenTelemetry .NET.

## Main settings

The following settings are common to most instrumentation scenarios:

| Setting                   | Description                                                                           | OpenTelemetry equivalent                       |
|---------------------------|---------------------------------------------------------------------------------------|------------------------------------------------|

> | `SIGNALFX_ENV`            | The value for the `deployment.environment` tag added to every span.                   | Should be set using `OTEL_RESOURCE_ATTRIBUTES` |

Reviewer 3: ```suggestion
| `SIGNALFX_ENV`            | The value for the `deployment.environment` tag added to every span.                   | Use `OTEL_RESOURCE_ATTRIBUTES` |
```

| `SIGNALFX_SERVICE_NAME`   | The name of the application or service.                                               | `OTEL_SERVICE_NAME`                            |

> | `SIGNALFX_VERSION`        | The version of the application. When set, it populates the `version` tag on spans.    | Should be set using `OTEL_RESOURCE_ATTRIBUTES` |

Reviewer 3: ```suggestion
| `SIGNALFX_VERSION`        | The version of the application. When set, it populates the `version` tag on spans.    |  Use `OTEL_RESOURCE_ATTRIBUTES` |
```

Reviewer 3: Can we add an example for both?

Author: added in https://github.com/signalfx/splunk-otel-dotnet/pull/124/commits/ab38c08fb5198569a45b5296a35b1b9d9cbf0688
